plugin/plugin.lua

```diff
@@ -0,0 +1,268 @@
+                    if exposedHeaders then
+                        ngx.header['access-control-expose-headers'] = exposedHeaders
+                    end
+                end
+                
+                if config.cors_max_age then
+                    if config.cors_max_age > 0 then
+                        ngx.header['access-control-max-age'] = config.cors_max_age
+                    end
+                end
+            end
+        end
+    end
+end
+
+local function error_response(status, code, message, config)
+
+    add_cors_response_headers(config)
+
+    local method = ngx.req.get_method():upper()
+    if method ~= 'HEAD' then
+    
+        local jsonData = '{"code":"' .. code .. '", "message":"' .. message .. '"}'
+        ngx.status = status
+        ngx.header['content-type'] = 'application/json'
+        ngx.say(jsonData)
+    end
+    
+    ngx.exit(status)
+end
+
+local function unauthorized_request_error_response(config)
+    error_response(ngx.HTTP_UNAUTHORIZED, 'unauthorized', 'Access denied due to missing or invalid credentials', config)
+end
+
+local function decrypt_cookie(encrypted_cookie, encryption_key_hex)
+
+    local all_bytes, err = base64.decode_base64url(encrypted_cookie)
+    if err then
+        ngx.log(ngx.WARN, 'A received cookie could not be base64url decoded ' .. err)
+        return nil
+    end
+
+    local min_size = VERSION_SIZE + GCM_IV_SIZE + 1 + GCM_TAG_SIZE
+    if #all_bytes < min_size then
+        ngx.log(ngx.WARN, 'A received cookie had an invalid length')
+        return nil
+    end
+
+    local offset = 1
+    local version_byte = string.byte(all_bytes, offset, VERSION_SIZE)
+    if version_byte ~= CURRENT_VERSION then
+        ngx.log(ngx.WARN, 'A received cookie had invalid format')
+        return nil
+    end
+
+    offset = 1 + VERSION_SIZE
+    local iv_bytes = string.sub(all_bytes, offset, VERSION_SIZE + GCM_IV_SIZE)
+  
+    offset = 1 + VERSION_SIZE + GCM_IV_SIZE
+    local ciphertext_bytes = string.sub(all_bytes, offset, #all_bytes - GCM_TAG_SIZE)
+
+    offset = #all_bytes - GCM_TAG_SIZE + 1
+    local tag_bytes = string.sub(all_bytes, offset)
+
+    local cipher = cipher.new('aes-256-gcm')
+    local encryption_key_bytes = from_hex(encryption_key_hex)
+    local decrypted_cookie, err = cipher:decrypt(encryption_key_bytes, iv_bytes, ciphertext_bytes, true, nil, tag_bytes)
+    if err then
+        ngx.log(ngx.WARN, 'Error decrypting cookie: ' .. err)
+        return nil
+    end
+
+    return decrypted_cookie
+end
+
+--
+-- The public entry point to decrypt a secure cookie from SPAs and forward the contained access token
+--
+function _M.run(config)
+
+    apply_configuration_defaults(config)
+
+    -- Pre-flight requests cannot contain cookies, so add CORS headers and return
+    local method = ngx.req.get_method():upper()
+    if method == 'OPTIONS' then
+        if config.cors_enabled then
+            add_cors_response_headers(config)
+        end
+        ngx.exit(200)
```

Comment:
Shouldn't there be `return` statements whenever we finish the response early - so here, and where error are returned?

Reply:
Done